File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/java/io/github/lishangbu/avalon/dataset/service/impl/BerryServiceImplTest.java
package io.github.lishangbu.avalon.dataset.service.impl;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.Berry;
import io.github.lishangbu.avalon.dataset.repository.BerryRepository;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class BerryServiceImplTest {
    @Mock private BerryRepository berryRepository;
    @InjectMocks private BerryServiceImpl berryService;

    @Test
    void getPageByCondition_callsRepository() {
        Berry berry = new Berry();
        Pageable pageable = PageRequest.of(0, 5);
        Page<Berry> page = new PageImpl<>(List.of(berry));
        when(berryRepository.findAll(any(Example.class), eq(pageable))).thenReturn(page);

        Page<Berry> result = berryService.getPageByCondition(berry, pageable);

        assertSame(page, result);
        ArgumentCaptor<Example> captor = ArgumentCaptor.forClass(Example.class);
        verify(berryRepository).findAll(captor.capture(), eq(pageable));
        assertNotNull(captor.getValue().getProbe());
    }

    @Test
    void save_usesRepository() {
        Berry berry = new Berry();
        when(berryRepository.save(berry)).thenReturn(berry);

        Berry result = berryService.save(berry);

        assertSame(berry, result);
        verify(berryRepository).save(berry);
    }

    @Test
    void update_usesRepository() {
        Berry berry = new Berry();
        when(berryRepository.save(berry)).thenReturn(berry);

        Berry result = berryService.update(berry);

        assertSame(berry, result);
        verify(berryRepository).save(berry);
    }

    @Test
    void removeById_callsRepository() {
        berryService.removeById(1L);
        verify(berryRepository).deleteById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/kotlin/io/github/lishangbu/avalon/dataset/service/impl/BerryServiceImplTest.kt
package io.github.lishangbu.avalon.dataset.service.impl

import io.github.lishangbu.avalon.dataset.entity.Berry
import io.github.lishangbu.avalon.dataset.entity.BerryFirmness
import io.github.lishangbu.avalon.dataset.entity.Type
import io.github.lishangbu.avalon.dataset.entity.dto.BerrySpecification
import io.github.lishangbu.avalon.dataset.entity.dto.BerryView
import io.github.lishangbu.avalon.dataset.entity.dto.SaveBerryInput
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateBerryInput
import io.github.lishangbu.avalon.dataset.repository.BerryRepository
import org.babyfish.jimmer.Page
import org.babyfish.jimmer.sql.ast.mutation.AssociatedSaveMode
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.springframework.data.domain.PageRequest

class BerryServiceImplTest {
    private val repository = mock(BerryRepository::class.java)
    private val service = BerryServiceImpl(repository)

    @Test
    fun getPageByCondition_callsRepository() {
        val specification = BerrySpecification(id = "1", internalName = "cheri")
        val pageable = PageRequest.of(0, 5)
        `when`(repository.pageViews(specification, pageable)).thenReturn(Page(listOf(berryView(1L)), 1, 1))

        val result = service.getPageByCondition(specification, pageable)

        assertEquals(1, result.rows.size)
        assertEquals("1", result.rows.first().id)
        assertEquals(
            "hard",
            result.rows
                .first()
                .berryFirmness
                ?.internalName,
        )
    }

    @Test
    fun save_usesInsertOnlyModeAndReloadsView() {
        `when`(repository.save(any<Berry>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(berrySavedEntity(1L))
        `when`(repository.loadViewById(1L)).thenReturn(berryView(1L))

        val result = service.save(SaveBerryInput("cheri", "樱子", 2, 3, 4, 5, 6, "7", "8", 9))

        assertEquals("1", result.id)
        assertEquals("hard", result.berryFirmness?.internalName)
        verify(repository).save(any<Berry>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())
        verify(repository).loadViewById(1L)
    }

    @Test
    fun update_usesUpsertModeAndReloadsView() {
        `when`(repository.save(any<Berry>(), eq(SaveMode.UPDATE_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(berrySavedEntity(1L))
        `when`(repository.loadViewById(1L)).thenReturn(berryView(1L))

        val result = service.update(UpdateBerryInput("1", "cheri", "樱子", 2, 3, 4, 5, 6, "7", "8", 9))

        assertEquals("1", result.id)
        assertEquals("硬", result.berryFirmness?.name)
        verify(repository).save(any<Berry>(), eq(SaveMode.UPDATE_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())
        verify(repository).loadViewById(1L)
    }

    @Test
    fun removeById_callsRepository() {
        service.removeById(1L)

        verify(repository).deleteById(1L)
    }
}

private fun berrySavedEntity(id: Long): Berry =
    Berry {
        this.id = id
        internalName = "cheri"
        name = "樱子"
        growthTime = 2
        maxHarvest = 3
        bulk = 4
        smoothness = 5
        soilDryness = 6
        berryFirmness =
            BerryFirmness {
                this.id = 7L
            }
        naturalGiftType =
            Type {
                this.id = 8L
            }
        naturalGiftPower = 9
    }

private fun berryWithAssociations(id: Long): Berry =
    Berry {
        this.id = id
        internalName = "cheri"
        name = "樱子"
        growthTime = 2
        maxHarvest = 3
        bulk = 4
        smoothness = 5
        soilDryness = 6
        berryFirmness =
            BerryFirmness {
                this.id = 7L
                internalName = "hard"
                name = "硬"
            }
        naturalGiftType =
            Type {
                this.id = 8L
                internalName = "fire"
                name = "火"
            }
        naturalGiftPower = 9
    }

private fun berryView(id: Long): BerryView = BerryView(berryWithAssociations(id))
